proposals/0010-diet-clients.txt

```diff
@@ -0,0 +1,183 @@
+to the bankhash that was signed by all the validators.
+
+#### StateHash: Transaction state transition merkle tree
+
+Merkle root of all the outputs from all the transactions and the
+status codes.
+
+### Future work: Separating execution and fork choice
+
+Eventually the solana protocol will separate fork choice from
+computing all the state transitions for all the programs. Diet
+clients would need to be modified to provide confirmation of execution
+for users.
+
+BankHash = (ShredHash, BlockHash, VoteStateHash, Prev BankHash)
+
+VoteStateHash = Current TX State Results for votes only
+
+ProgramStateHash = (Current TX State Results, Previous ProgramStateHash)
+
+While validators vote only on the BankHash, users still need to
+know the ProgramStateHash for the slot. 
+
+Validators are not expected to compute the ProgramStateHash at the
+same time as the BankHash, and the ProgramStateHash may lag.  But
+at least once an epoch, all the validators must compute the EpochHash
+for the previous epoch, which includes all the ProgramStateHashe from the
+previous epoch.
+
+Epoch Hash = Merkle root of all the accounts including the account
+that stores ProgramStateHashes for the epoch
+
+#### Overview
+
+The original protocol is modified as follows:
+
+2. Concurrently, user requests at random for votes from N different
+nodes. Votes maybe for later slots, and would contain a path to the
+requested slot's bankhash.
+2.a User also requests signed ProgramStateHash attestations from
+the staked validators.
+
+If the validator signed an invalid ProgramStateHash, this signed
+attestation can be used to slash them after the EpochHash has been
+computed by the network.
+
+### Future work: Minimized fraud proof
+
+Currently, the above proposal requires a client to validate the
+entire chain starting from a previously trusted snapshot up to the
+invalid block. This proposal covers a minimized fraud proof that
+allows the user to detect when the majority signed an invalid state
+transition. User would only need enough resources to process the
+votes and one transaction.
+
+#### Overview
+
+1. Track all the inputs into state transitions
+
+Each input should have a merkle path to the original snapshot, or
+if it has been modified, to an output from a previous transaction.
+
+2. Merkelize all the txs and their inputs and outputs for each block
+
+       StateHash
+     h1       h2
+
+  tx1  tx2  tx3 tx4
+
+tx:
+    inputs: all read account paths: path to snapshot | path to previous
+            block output | path to intra-block tx output
+    result: success or failure code
+    outputs: all write accounts if result is successful, 0 otherwise
+
+The tree should contain an ordered sequence of txs. Each tx is its
+own tree that contains all the input paths to the accounts used for
+execution and all the outputs. Each input can either be an
+output from a snapshot root, or an output from a previous block,
+or from a transaction in the same block.
```

Comment:
Should the user also be checking some non-inclusion proofs, to make sure no transaction has modified the input in-between? For example, if the user is given the input that is an output from a snapshot root, should they confirm there were no other transactions modifying the input between the snapshot and the current transaction?

Reply:
@valerini, An honest validator would have computed a different input using a more recent value and path for that TX, so the result would have been different.  The honest validator can then return the more recent path and the majority signed header as the proof.